File: ruby/triangle/triangle.rb
class Triangle
  IS_EQUILATERAL = ->(unique_sides) { unique_sides == 1 }
  private_constant :IS_EQUILATERAL
  IS_ISOCELES = ->(unique_sides) { unique_sides < 3 }
  private_constant :IS_ISOCELES
  IS_SCALENE = ->(unique_sides) { unique_sides == 3 }
  private_constant :IS_SCALENE

  def initialize(sides)
    @sides = sides.sort
  end

  def equilateral?
    legal_triangle?(&IS_EQUILATERAL)
  end

  def isosceles?
    legal_triangle?(&IS_ISOCELES)
  end

  def scalene?
    legal_triangle?(&IS_SCALENE)
  end

  private

  attr_reader :sides

  def legal_triangle?(&condition)
    sides.all?(&:positive?) &&
      side_lengths_all_legal? &&
      sides.uniq.size.yield_self(&condition)
  end

  def side_lengths_all_legal?
    sides.yield_self do |(*first_two_sides, third_side)|
      first_two_sides.sum >= third_side
    end
  end
end

File: typescript/triangle/triangle.ts
type MaybeError<T> = T | never
type Sides = [number, number, number]
type TriangleSides = Readonly<Sides>

export default class Triangle {
  private readonly IS_EQUILATERAL = (uniqueSides: number[]): boolean => {
    return uniqueSides.length === 1
  }
  private readonly IS_ISOCELES = (uniqueSides: number[]): boolean => {
    return uniqueSides.length < 3
  }
  private readonly IS_SCALENE = (uniqueSides: number[]): boolean => {
    return uniqueSides.length === 3
  }
  private readonly sides: TriangleSides

  constructor(...sides: Sides) {
    this.sides = sides.sort(this.ascending)
  }

  kind(): MaybeError<string> {
    if (!this.allSidesLegal()) {
      throw new Error("Invalid triangle sides")
    }

    const uniqueSides: number[] = this.uniqueSides()

    if (this.IS_EQUILATERAL(uniqueSides)) {
      return "equilateral"
    } else if (this.IS_ISOCELES(uniqueSides)) {
      return "isosceles"
    } else if (this.IS_SCALENE(uniqueSides)) {
      return "scalene"
    } else {
      throw new Error("Unexpected triangle type")
    }
  }

  private ascending(a: number, b: number): number {
    return a - b
  }

  private uniqueSides(): number[] {
    return [...new Set(this.sides)]
  }

  private allSidesLegal(): boolean {
    return this.allSidesPositive() && this.allSidesLegalLength()
  }

  private allSidesPositive(): boolean {
    return this.sides.every(side => side > 0)
  }

  private allSidesLegalLength(): boolean {
    const [firstSide, secondSide, thirdSide]: TriangleSides = this.sides
    return firstSide + secondSide >= thirdSide
  }
}
